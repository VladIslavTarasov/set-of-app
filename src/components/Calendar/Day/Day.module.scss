.button {
  position: relative;
  width: 100%;
  padding: var(--offset-md) 0;
  border: none;
  background-color: transparent;
  color: var(--gray);
  font-size: var(--fs-lg);
  text-align: center;

  &::after {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 100%;
    height: 100%;
    background-color: transparent;
    border-radius: var(--radius-md);
    content: '';
    transform: translate(-50%, -50%) scale(0.7);
    transition: var(--transition);
  }

  &:focus,
  &:hover {
    &::after {
      background-color: var(--black1);
      opacity: 1;
    }
  }

  &:hover {
    &::after {
      animation: buttonAnimation 4s infinite;
      cursor: pointer;
    }
  }

  &:disabled {
    &::after {
      cursor: wait;
    }
  }
}

.activeMonth {
  color: var(--black);
}

.today {
  &::after {
    border: 2px solid var(--black1);
    background-color: var(--green2);
    transform: translate(-50%, -50%) scale(0.8);
  }
}

.choosen {
  &::after {
    background-color: var(--gray1);
    transform: translate(-50%, -50%) scale(0.8);
  }

  &:hover,
  &:focus {
    &::after {
      border: 2px solid var(--black2);
      animation: none;
      background-color: var(--gray1);
      transform: translate(-50%, -50%) scale(0.8);
    }
  }
}

@keyframes buttonAnimation {
  0% {
    transform: translate(-50%, -50%) scale(0.8);
  }

  50% {
    transform: translate(-50%, -50%) scale(1);
  }

  100% {
    transform: translate(-50%, -50%) scale(0.8);
  }
}
