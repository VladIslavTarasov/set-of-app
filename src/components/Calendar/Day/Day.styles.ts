import { createUseStyles } from 'react-jss';

import { Theme } from 'theme/types/theme.types';

export const useStyles = createUseStyles(
  {
    '@keyframes buttonAnimation': {
      '0%': {
        transform: 'translate(-50%, -50%) scale(0.8)',
      },

      '50%': {
        transform: 'translate(-50%, -50%) scale(0.9)',
      },

      '100%': {
        transform: 'translate(-50%, -50%) scale(0.8)',
      },
    },
    button: ({ theme }: { theme: Theme }) => ({
      position: 'relative',
      width: 60,
      height: 60,
      padding: `${theme.offsets.md} 0px`,
      border: 'none',
      backgroundColor: 'transparent',
      color: theme.palette.common.gray,
      fontSize: theme.fontSize.lg,
      textAlign: 'center',

      '& > span': {
        position: 'relative',
        zIndex: 2,
      },

      '&::after': {
        content: "''",
        position: 'absolute',
        zIndex: 1,
        top: '50%',
        left: '50%',
        width: '100%',
        height: '100%',
        backgroundColor: 'transparent',
        borderRadius: theme.borderRadius.md,
        transform: 'translate(-50%, -50%) scale(0.7)',
        transition: theme.transition.xs,
      },

      '&:focus, &:hover': {
        color: theme.palette.common.white,
        '&::after': {
          borderRadius: theme.borderRadius.md2,
          backgroundColor: theme.palette.primary.light,
        },
      },

      '&:hover': {
        '&::after': {
          animation: '$buttonAnimation 4s infinite',
          cursor: 'pointer',
        },
      },

      '&:disabled': {
        '&::after': {
          cursor: 'wait',
        },
      },
    }),
    activeMonth: ({ theme }: { theme: Theme }) => ({
      color: theme.palette.common.black,
    }),
    today: ({ theme }: { theme: Theme }) => ({
      color: theme.palette.common.white,

      '&::after': {
        border: `1px solid ${theme.palette.common.gray}`,
        borderRadius: theme.borderRadius.md2,
        backgroundColor: theme.palette.primary.light,
        transform: 'translate(-50%, -50%) scale(0.8)',
      },
    }),
    choosen: ({ theme }: { theme: Theme }) => ({
      color: theme.palette.common.white,

      '&::after': {
        backgroundColor: theme.palette.primary.main,
        borderRadius: theme.borderRadius.md2,
        transform: 'translate(-50%, -50%) scale(0.8)',
      },

      '&:hover, &:focus': {
        '&::after': {
          boxSizing: 'border-box',
          border: `1px solid ${theme.palette.common.gray}`,
          animation: 'none',
          backgroundColor: theme.palette.primary.main,
          transform: 'translate(-50%, -50%) scale(0.8)',
        },
      },
    }),
  },
  { name: 'Day' }
);
