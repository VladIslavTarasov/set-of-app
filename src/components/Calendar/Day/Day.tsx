import React, { memo, useCallback, useContext } from 'react';

import cn from 'classnames';
import moment from 'moment';

import { CalendarDispatch } from 'context/calendarDispatch';
import * as calendarActions from 'store/calendar/calendar.actions';

import style from './Day.module.scss';

const now = moment().unix();

const methods = {
  next: calendarActions.toNextMonth,
  prev: calendarActions.toPrevMonth,
};

interface DayProps {
  onChangeDate: (date: string) => void;
  day: number;
  date: string;
  choosenDate: string;
  isActiveMonth: boolean;
  isToday?: boolean;
  loading?: boolean;
  action?: 'next' | 'prev';
}

const Day: React.FC<DayProps> = ({
  onChangeDate,
  isActiveMonth,
  date,
  choosenDate,
  day,
  loading,
  isToday,
  action,
}) => {
  const dispatch = useContext(CalendarDispatch);

  const handleClick = useCallback(() => {
    if (action) {
      dispatch(methods[action]());
    }

    onChangeDate(date);
  }, [date, action, dispatch, onChangeDate]);

  return (
    <button
      type="button"
      onClick={handleClick}
      title={date}
      disabled={loading}
      className={cn(style.button, {
        [style.activeMonth]: isActiveMonth,
        [style.today]: isToday,
        [style.past]: moment(new Date(date)).unix() <= now && !isToday,
        [style.choosen]: choosenDate === date,
        [style.disabled]: loading,
      })}
    >
      {day}
    </button>
  );
};

export default memo(Day);
