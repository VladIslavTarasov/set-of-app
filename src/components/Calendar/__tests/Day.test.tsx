import React from 'react';

import { render } from '@testing-library/react';

import Day from '../Day';

describe('components/Calendar', () => {
  it('/Day', () => {
    const dayProps = {
      // eslint-disable-next-line
      onChangeDate: (date: string) => {},
      isActiveMonth: true,
      date: '2020-05-5',
      choosenDate: '2020-05-7',
      day: 5,
    };

    const { container, rerender } = render(<Day {...dayProps} />);
    expect(container).toMatchSnapshot();

    rerender(<Day {...dayProps} isToday />);
    expect(container).toMatchSnapshot();

    rerender(<Day {...dayProps} isToday loading />);
    expect(container).toMatchSnapshot();

    rerender(<Day {...dayProps} isToday loading choosenDate={dayProps.date} />);
    expect(container).toMatchSnapshot();
  });
});
