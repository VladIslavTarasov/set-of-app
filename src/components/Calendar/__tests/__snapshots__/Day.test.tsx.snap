// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Calendar /Day 1`] = `
<div>
  <button
    class="button activeMonth past"
    title="2020-05-5"
    type="button"
  >
    5
  </button>
</div>
`;

exports[`components/Calendar /Day 2`] = `
<div>
  <button
    class="button activeMonth today"
    title="2020-05-5"
    type="button"
  >
    5
  </button>
</div>
`;

exports[`components/Calendar /Day 3`] = `
<div>
  <button
    class="button activeMonth today disabled"
    disabled=""
    title="2020-05-5"
    type="button"
  >
    5
  </button>
</div>
`;

exports[`components/Calendar /Day 4`] = `
<div>
  <button
    class="button activeMonth today choosen disabled"
    disabled=""
    title="2020-05-5"
    type="button"
  >
    5
  </button>
</div>
`;
